static/js/index.js

```diff
@@ -57,77 +60,52 @@ function getXYOffsetOfRep(selStart, selEnd){
+  div.css(divEl.offset());
   var divWidth = div.width();
 
   // Is the line visible yet?
-  if ( div.length !== 0 ) {
-    var top = $(div).offset().top -10; // A standard generic offset
-
+  if ( divEl.length !== 0 ) {
+    var top = divEl.offset().top + topCorrection; // A standard generic offset
     // Get the HTML
     var html = $(div).html();
-
-    // if Div contains block attribute IE h1 or H2 then increment by the number
-    if ( $(div).children("span").length < 1 ){ x = x - 1; }// This is horrible but a limitation because I'm parsing HTML
-
-    // Get the new string but maintain mark up
-    var newText = html_substr(html, (x));
-
-    // A load of fugly HTML that can prolly be moved ot CSS
-    var newLine = "<span style='width:"+divWidth+"px' id='hiddenWorker' class='ghettoCursorXPos'>"+newText+"</span>";
-    // GlobalKey
-    globalKey = 0;
-
-    // Add the HTML to the DOM
-    var worker = $('iframe[name="ace_outer"]').contents().find('#outerdocbody').append(newLine);
-
-    // Get the worker element
-    var worker = $('iframe[name="ace_outer"]').contents().find('#outerdocbody').find("#hiddenWorker");
-
-    // add some CSS to the worker
-    $(worker).css({
-      "white-space":"pre-wrap",
-      "word-wrap":"break-word",
-      "z-index":"99999",
-      "background":"red",
-      "position":"fixed",
-      "top":"80px",
-      "left":"33px",
-      "margin-right":"10px",
-      "font-size":"12px",
-      "line-height":"16px"
-    })
-
-    // Wrap teh HTML in spans so we cna find a char
-    $(worker).html(wrap($(worker), true));
-    // console.log($(worker).html(), x);
-
-    // Get the Left offset of the x span
-    var span = $(worker).find("[data-key="+(x-1)+"]");
-
-    // Get the width of the element (This is how far out X is in px);
-    if(span.length !== 0){
-      var left = span.position().left;
-    }else{
-      var left = 0;
+    var text = $(div).text().split('');
+    var workerIndex;
+    if (viewPosition === 'right') {
+      if (selEnd[0] > selStart[0] && selEnd[1] === 0) {
+        line =  getLineAtIndex(selEnd[0] -1);
+        workerIndex = $(line.lineNode).text().length - 1;
+      } else {
+        workerIndex = selEnd[1];
+      }
+    } else {
+      workerIndex = selStart[1];
     }
-
-    // Get the height of the element minus the inner line height
-    var height = worker.height(); // the height of the worker
-    top = top + height - span.height(); // plus the top offset minus the actual height of our focus span
-
-    if(top <= 0){  // If the tooltip wont be visible to the user because it's too high up
-      stickUp = true;
-      top = top + (span.height()*2);
-      if(top < 0){ top = 0; } // handle case where caret is in 0,0
+    text.splice(workerIndex, 0, '</span>');
+    text.splice(workerIndex, 0, '<span id="selectWorker">');
+    var heading = isHeading();
+    text = text.join('');
+    if (heading) {
+      text = '<'+heading+'>' + text + '</'+heading+'>';
```

Comment:
Shouldn't we escape this string for security reasons?

Reply:
Good, notice. For some reason I thought that if text was already in pad, it's escaped. But Now I have corrected this.